File: src/SharpOMatic.Engine/DTO/WorkflowRunMetricsDashboard.cs
namespace SharpOMatic.Engine.DTO;

public enum WorkflowRunMetricScope
{
    All = 0,
    Workflow = 1,
    Error = 2,
}

public sealed record WorkflowRunMetricsDashboardRequest(
    DateTime Start,
    DateTime End,
    ModelCallMetricBucket Bucket,
    WorkflowRunMetricScope Scope,
    string? ScopeKey,
    string? MasterSearch,
    int RecentSkip,
    int RecentTake,
    bool AllTime = false
);

public sealed record WorkflowRunMetricsDashboard(
    DateTime Start,
    DateTime End,
    ModelCallMetricBucket Bucket,
    WorkflowRunMetricScope Scope,
    string? ScopeKey,
    string? ScopeName,
    WorkflowRunMetricTotals Totals,
    List<WorkflowRunMetricMasterItem> MasterItems,
    List<WorkflowRunMetricTimeBucket> TimeBuckets,
    List<WorkflowRunMetricBreakdownItem> WorkflowBreakdown,
    List<WorkflowRunMetricFailureGroup> Failures,
    List<WorkflowRunMetricRunSummary> RecentRuns,
    int RecentRunsTotal,
    List<WorkflowRunMetricRunSummary> SlowestRuns
);

public sealed record WorkflowRunMetricTotals(
    int TotalRuns,
    int SuccessfulRuns,
    int FailedRuns,
    int SuspendedRuns,
    double? AverageDuration,
    long? P95Duration,
    double FailureRate,
    int ModelCallCount,
    int ModelCallFailureCount,
    long InputTokens,
    long OutputTokens,
    long TotalTokens,
    decimal TotalModelCost
);

public sealed record WorkflowRunMetricMasterItem(
    string Key,
    string Name,
    int TotalRuns,
    int SuccessfulRuns,
    int FailedRuns,
    int SuspendedRuns,
    double? AverageDuration,
    long? P95Duration,
    double FailureRate,
    int ModelCallCount,
    int ModelCallFailureCount,
    long TotalTokens,
    decimal TotalModelCost
);

public sealed record WorkflowRunMetricTimeBucket(
    DateTime Start,
    int TotalRuns,
    int SuccessfulRuns,
    int FailedRuns,
    int SuspendedRuns,
    double? AverageDuration,
    long? P95Duration,
    int ModelCallCount,
    int ModelCallFailureCount,
    long InputTokens,
    long OutputTokens,
    long TotalTokens,
    decimal TotalModelCost
);

public sealed record WorkflowRunMetricBreakdownItem(
    string Key,
    string Name,
    int TotalRuns,
    int SuccessfulRuns,
    int FailedRuns,
    int SuspendedRuns,
    double? AverageDuration,
    long? P95Duration,
    double FailureRate,
    int ModelCallCount,
    int ModelCallFailureCount,
    long TotalTokens,
    decimal TotalModelCost
);

public sealed record WorkflowRunMetricFailureGroup(
    string ErrorType,
    string ErrorMessage,
    Guid? FailedNodeEntityId,
    string FailedNodeTitle,
    NodeType? FailedNodeType,
    int Count,
    DateTime FirstSeen,
    DateTime LastSeen,
    int WorkflowCount
);

public sealed record WorkflowRunMetricRunSummary(
    Guid Id,
    Guid RunId,
    DateTime Created,
    DateTime Started,
    DateTime Finished,
    long? Duration,
    Guid WorkflowId,
    string WorkflowName,
    int WorkflowVersion,
    bool Succeeded,
    RunStatus RunStatus,
    string? ErrorType,
    string? ErrorMessage,
    Guid? FailedNodeEntityId,
    string? FailedNodeTitle,
    NodeType? FailedNodeType,
    string? ConversationId,
    int? TurnNumber,
    bool IsConversationRun,
    int ModelCallCount,
    int ModelCallFailureCount,
    long InputTokens,
    long OutputTokens,
    long TotalTokens,
    decimal TotalModelCost
);

File: src/SharpOMatic.FrontEnd/src/app/pages/metrics/interfaces/workflow-run-metrics-dashboard.ts
import { RunStatus } from '../../../enumerations/run-status';
import { NodeType } from '../../../entities/enumerations/node-type';
import { ModelCallMetricBucket } from './model-call-metrics-dashboard';

export enum WorkflowRunMetricScope {
  All = 0,
  Workflow = 1,
  Error = 2,
}

export interface WorkflowRunMetricsDashboard {
  start: string;
  end: string;
  bucket: ModelCallMetricBucket;
  scope: WorkflowRunMetricScope;
  scopeKey: string | null;
  scopeName: string | null;
  totals: WorkflowRunMetricTotals;
  masterItems: WorkflowRunMetricMasterItem[];
  timeBuckets: WorkflowRunMetricTimeBucket[];
  workflowBreakdown: WorkflowRunMetricBreakdownItem[];
  failures: WorkflowRunMetricFailureGroup[];
  recentRuns: WorkflowRunMetricRunSummary[];
  recentRunsTotal: number;
  slowestRuns: WorkflowRunMetricRunSummary[];
}

export interface WorkflowRunMetricTotals {
  totalRuns: number;
  successfulRuns: number;
  failedRuns: number;
  suspendedRuns: number;
  averageDuration: number | null;
  p95Duration: number | null;
  failureRate: number;
  modelCallCount: number;
  modelCallFailureCount: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  totalModelCost: number;
}

export interface WorkflowRunMetricMasterItem {
  key: string;
  name: string;
  totalRuns: number;
  successfulRuns: number;
  failedRuns: number;
  suspendedRuns: number;
  averageDuration: number | null;
  p95Duration: number | null;
  failureRate: number;
  modelCallCount: number;
  modelCallFailureCount: number;
  totalTokens: number;
  totalModelCost: number;
}

export interface WorkflowRunMetricTimeBucket {
  start: string;
  totalRuns: number;
  successfulRuns: number;
  failedRuns: number;
  suspendedRuns: number;
  averageDuration: number | null;
  p95Duration: number | null;
  modelCallCount: number;
  modelCallFailureCount: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  totalModelCost: number;
}

export interface WorkflowRunMetricBreakdownItem {
  key: string;
  name: string;
  totalRuns: number;
  successfulRuns: number;
  failedRuns: number;
  suspendedRuns: number;
  averageDuration: number | null;
  p95Duration: number | null;
  failureRate: number;
  modelCallCount: number;
  modelCallFailureCount: number;
  totalTokens: number;
  totalModelCost: number;
}

export interface WorkflowRunMetricFailureGroup {
  errorType: string;
  errorMessage: string;
  failedNodeEntityId: string | null;
  failedNodeTitle: string;
  failedNodeType: NodeType | null;
  count: number;
  firstSeen: string;
  lastSeen: string;
  workflowCount: number;
}

export interface WorkflowRunMetricRunSummary {
  id: string;
  runId: string;
  created: string;
  started: string;
  finished: string;
  duration: number | null;
  workflowId: string;
  workflowName: string;
  workflowVersion: number;
  succeeded: boolean;
  runStatus: RunStatus;
  errorType: string | null;
  errorMessage: string | null;
  failedNodeEntityId: string | null;
  failedNodeTitle: string | null;
  failedNodeType: NodeType | null;
  conversationId: string | null;
  turnNumber: number | null;
  isConversationRun: boolean;
  modelCallCount: number;
  modelCallFailureCount: number;
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  totalModelCost: number;
}
